src/manifold.jl

```diff
@@ -33,30 +20,34 @@ properties.
 
 ## Arguments
 
-  - `g`: The residual function for the manifold.
-
-      * This is an inplace function of form `g(resid, u, p)` or `g(resid, u, p, t)` which
-        writes to the residual `resid` the difference from the manifold components. Here, it
-        is assumed that `resid` is of the same shape as `u`.
-      * If `isinplace = Val(false)`, then `g` should be a function of the form `g(u, p)` or
-        `g(u, p, t)` which returns the residual.
+  - `manifold`: The residual function for the manifold. If the ODEProblem is an inplace
+    problem, then `g` must be an inplace function of form `g(resid, u, p)` or
+    `g(resid, u, p, t)` and similarly if the ODEProblem is an out-of-place problem then `g`
+    must be a function of form `g(u, p)` or `g(u, p, t)`.
 
 ## Keyword Arguments
 
-  - `nlsolve`: A nonlinear solver as defined in the
-    [NonlinearSolve.jl format](https://docs.sciml.ai/NonlinearSolve/stable/basics/solve/).
-    Defaults to a PolyAlgorithm.
+  - `nlsolve`: Defaults to a special single-factorize algorithm (denoted by `missing`) that
+    would work in most cases (See [1] for details). Alternatively, a nonlinear solver as
+    defined in the
+    [NonlinearSolve.jl format](https://docs.sciml.ai/NonlinearSolve/stable/basics/solve/)
+    can be specified.
   - `save`: Whether to do the standard saving (applied after the callback)
-  - `nlls`: If the problem is a nonlinear least squares problem. `nlls = Val(false)`
-    generates a `NonlinearProblem` which is typically faster than
-    `NonlinearLeastSquaresProblem`, but is only applicable if the residual size is same as
-    the state size.
   - `autonomous`: Whether `g` is an autonomous function of the form `g(resid, u, p)` or
-    `g(u, p)`. Specify it as `Val(::Bool)` to ensure this function call is type stable.
-  - `residual_prototype`: This needs to be specified if `nlls = Val(true)` and
-    `inplace = Val(true)` are specified together, else it is taken to be same as `u`.
+    `g(u, p)`. Specify it as `Val(::Bool)` to disable runtime branching. If `nothing`,
+    we attempt to infer it from the signature of `g`.
+  - `residual_prototype`: The size of the manifold residual. If it is not specified,
+    we assume it to be same as `u` in the inplace case. Else we run a single evaluation of
+    `manifold` to determine the size.
   - `kwargs`: All other keyword arguments are passed to
-    [NonlinearSolve.jl](https://docs.sciml.ai/NonlinearSolve/stable/basics/solve/).
+    [NonlinearSolve.jl](https://docs.sciml.ai/NonlinearSolve/stable/basics/solve/) if
+    `nlsolve` is not `missing`.
+  - `autodiff`: The autodifferentiation algorithm to use to compute the Jacobian if
```

Comment:
making this mandatory for the time being. Once NonlinearSolveBase is released, we can reconsider that.